ITP_LONDON_JAN_2025| CRISTINA_MARIA_LEASCU | DATA_GROUPS | Alarmclock

CristinaLeascu opened the pull request:
<!--

You must title your PR like this:

COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK

For example,

NW4 | Carol Owen | HTML-CSS-Module | Week1

Complete the task list below this message.
If your PR is rejected, check the task list.

-->

## Learners, PR Template

Self checklist

- [ ] I have committed my files one by one, on purpose, and for a reason
- [ ] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [ ] I have tested my changes
- [ ] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [ ] My changes meet the [requirements](./README.md) of this task

## Changelist

Briefly explain your PR.

## Questions

Ask any questions you have for your reviewer.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -23,3 +80,4 @@ function pauseAlarm() {
 }
 
 window.onload = setup;
+module.exports = setAlarm;
```
This statement would cause a runtime error when the app is running in a browser.

When you test your app in the browser, you can check the JS console in the browser to see if there is any runtime error message.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -1,4 +1,61 @@
+  const formattedMinutes = String(minutes).padStart(2, "0");
+  const formattedRemainingSeconds = String(remainingSeconds).padStart(2, "0");
+
+  console.log(
+    "formatted time (mm:ss):",
+    `${formattedMinutes}:${formattedRemainingSeconds}`
+  );
```
We should remove or comment out debugging statements in the production code.

CristinaLeascu replied:
Thank you for your feedback, log removed,

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -1,4 +1,61 @@
+    if (timeInSeconds <= 0) {
+      clearInterval(countdown); // Stop the timer
+
+      let flashCount = 0; // To track the number of flashes
+      const colors = ["pink", "plum", "indigo", "lavander blush"];
+      const flashInterval = setInterval(() => {
+        document.body.style.backgroundColor =
+          colors[flashCount % colors.length];
+        flashCount++;
+
+        if (flashCount >= colors.length) {
+          // Stop flashing after 3 flashes
+          clearInterval(flashInterval); // Stop flashing
+          document.body.style.backgroundColor = "pink"; // Keep background red
+          playAlarm();
+        }
+      }, 500);
+    }
+  }, 1000);
```
There isn't seem to be a way to reset the changed background to its initial state (white background).

Restarting a new countdown immediately after the countdown reaches 0 does not seem to be able to stop the blinking. I expect restarting a countdown would reset the appearance.

If you have time, you can consider using only CSS to animate the background and then use JS to toggle a CSS class to add/remove the effect.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -1,4 +1,61 @@
+    timeInSeconds--;
+
+    // Update the display each second
+    timeRemaining.textContent = `Time Remaining: ${formatTime(timeInSeconds)}`;
```
If we place the statement that displays the countdown clock (i.e., the code that update the view) in a function,
- we can replace the lines 28-29 by a function call, and
- in case we need to change the wordings or the way to display the time, we only need to change one set of code

CristinaLeascu replied:
Thank you. I have implemented a function.

cjyuan reviewed (commented):
Code works fine if a user only clicks the "Set Alarm" button once.
However, if the user enters a time and then clicks the "Set Alarm" button multiple times, the countdown clock will not display properly.

Also, the background color, once changed, cannot be reset to its initial color.

Can you fix the issues?

CristinaLeascu commented:
@cjyuan  Thank you for your feedback.
I have removed log, created functions for countdownClock, resetBackground, flashBackground and updated my code.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -2,63 +2,95 @@
 
-  const formattedMinutes = String(minutes).padStart(2, "0");
-  const formattedRemainingSeconds = String(remainingSeconds).padStart(2, "0");
+  window.flashInterval = setInterval(() => {
```
Why introduce `flashInterval` as a property of `window` object (which has the same effect as declaring it in the global scope using 'var') but declare `countdownInterval` at line 19 using `let`?

Both approaches work. I am just curious why you used two different approaches.

You can ask ChatGPT the difference between declaring global variables using `var` and `let`.

CristinaLeascu replied:
flashInterval controls the flashing of the background, would have to be accesible for other functions globally accesible. 
countdownInterval it was intended to be used for countdownClock only and the I have added to the setAlarm(). I understood now the diffrence. Thank you.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -2,63 +2,95 @@
+function countdownClock(timeInSeconds) {
+  const timeRemaining = document.querySelector("#timeRemaining");
+
+  if (countdownInterval) {
+    clearInterval(countdownInterval);
+  }
```
The if-statement is not needed. Can you figure out why we can just call `clearInterval(countdownInterval);` without the if-statement?

CristinaLeascu replied:
Thank you for your feedback, after reaserching I have understood that I can use clearInterval(countdownInterval);, as clearInterval it handles invalid interval gracefuly. I will stop the interval or ingnore the call.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -2,63 +2,95 @@
+      clearInterval(window.flashInterval);
+      document.body.style.backgroundColor = "white"; // set the final background color to white
+
+      setTimeout(() => {
+        playAlarm(); // start the alart af the end of backgroung flash
+      });
```
What's the difference between
```
      setTimeout(() => {
        playAlarm();
      });
```
and
```
        playAlarm();
```

Why not just use the latter approach?

CristinaLeascu replied:
I has some issues with the alarm starting when countdown reached 00:00 
and not after the background flash, I introduced 2000 delay to give time for flashing then removed and I left it as it worked.

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -2,63 +2,95 @@
-  );
-  return `${formattedMinutes}:${formattedRemainingSeconds}`;
+    if (flashCount >= colors.length) {
+      clearInterval(window.flashInterval);
+      document.body.style.backgroundColor = "white"; // set the final background color to white
```
Why not just call `resetBackground()`?

cjyuan commented on Sprint-3/alarmclock/alarmclock.js:
```diff
@@ -2,63 +2,95 @@
+}
+
+// function to handle countdown clock
+let countdownInterval = null;
```
Common best practice is to declare global variables at the top of the files; it makes it easier to see and manage them.

CristinaLeascu replied:
Thank yiu fir yiur feedback, after reaserching I have understood that I can use clearInterval(countdownInterval);, as clearInterval it handles invalid interval gracefuly. I will stop the interval or ingnore the call.
